### PR TITLE
`CODEOWNERS` and branches protection

### PR DESCRIPTION
**Pull request by xeho91**
# Code owners

**This Pull Request starts the discussion on how to organize the responsibilities delegation for this project.**

---

## TODO

- [x] Configure GitHub's `CODEOWNERS` file
- [ ] Document the decision in our documentation files or Wiki pages?

---

## Initial proposal

A code owner is a person responsible for the code review and required to accept the changes.

## Project Manager

@xeho91 

### Directories

- `client/`, aka **front-end** - @mariushilleke as a core code owner + @xeho91
- `server/`, aka **back-end** - @lax0248917 as a core code owner + @xeho91

### Branches

Conditions for managing the operations in the following branches
- `main` - requires **everyone's** approval (@xeho91, @mariushilleke, @lax0248917)
- `develop` requires **two** approval from the person responsible for the directory where the changes have been made and [Project Manager](#project-manager)

---

Thoughts?

### CODE REVIEW

**Review by lax0248917 — COMMENTED** [review]
This is a good start but I believe we discussed this a bit further in discord. Creating the requirement of two approvals for merging pull requests will help us make sure we don't approve our own code. And adds a layer of safety to the main branch.

Also, how we will we handle PRs that reference code that resides outside of these directories? (ex. README files)

**Comment by xeho91** [comment]
> This is a good start but I believe we discussed this a bit further in discord. Creating the requirement of two approvals for merging pull requests will help us make sure we don't approve our own code. And adds a layer of safety to the main branch.

Approvals are being set in the repository setting under the branch rules, not in the [`CODEOWNERS`] file. As far I know.
I already made these changes. Feel free to check.

> Also, how we will we handle PRs that reference code that resides outside of these directories? (ex. README files)

Good point. I did expand the ownership to include everything at the root directory, except specific folders.
See the latest commit 5b4e1d7.

[`CODEOWNERS`]: ./.github/CODEOWNERS

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1,6 +1,10 @@
 # Default owner for everything in the repo.
 * @xeho91
 
-# Directories - from the project root
+# Include everything in the repo - from the project root, EXCEPT...
+!client/ @lax0248917
+!server/ @mariushilleke
```
Comment (xeho91) [inline .github/CODEOWNERS:6]: @lax0248917 

Here. This should work.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -1,6 +1,10 @@
...
+
+# Include directories - from the project root
 /client/ @mariushilleke
 /server/ @lax0248917
```
Comment (xeho91) [inline .github/CODEOWNERS]: Technically, this is unnecessary now, but it makes it more "_clear_"? About who is the code owner for these directories. Should we keep it?

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Default owner for everything in the repo.
+* @xeho91
+
+# Include everything in the repo - from the project root, EXCEPT...
+!/client/ @lax0248917
+!/server/ @mariushilleke
```
Comment (xeho91) [inline .github/CODEOWNERS:6]: @lax0248917 

Here. The `!` is a negating pattern. To my knowledge, it means "_everything from the root, except..._".

Comment (lax0248917) [reply .github/CODEOWNERS:6]: We have discussed this in discord and are going to institute the use of teams to make sure the correct code submissions are being reviewed by the right people

**Comment by xeho91** [comment]
**QUESTION**: Any thoughts on documenting the current project team without the need to update the [README.md] file frequently?

There is a draft to add this to the [README.md] file:

---

## ☯️ Project team

[@xeho91] - Project Manager\
[@lax0248917] - core Back-End Leader\
[@mariushilleke] - core Front-End Leader

> 📝**NOTE:** We are flexible when it comes to roles. It may change over the
> time.

[@xeho91]: https://github.com/xeho91
[@lax0248917]: https://github.com/lax0248917
[@mariushilleke]: https://github.com/mariushilleke

---

[README.md]: ./README.md

**Pull request closed without merging**